src/pymmcore_plus/mda/handlers/_ome_tiff_writer.py

```diff
@@ -0,0 +1,107 @@
+
+    def sequenceStarted(self, seq: useq.MDASequence) -> None:
+        self._set_sequence(seq)
+
+    def frameReady(self, frame: np.ndarray, event: useq.MDAEvent, meta: dict) -> None:
+        if self._mmap is None:
+            if not self._current_sequence:
+                # just in case sequenceStarted wasn't called
+                self._set_sequence(event.sequence)  # pragma: no cover
+
+            if not (seq := self._current_sequence):
+                raise NotImplementedError(
+                    "Writing zarr without a MDASequence not yet implemented"
+                )
+
+            mmap = self._create_seq_memmap(frame, seq, meta)
+        else:
+            mmap = self._mmap
+
+        # WRITE DATA TO DISK
+        index = tuple(event.index.get(k) for k in self._used_axes)
+        mmap[index] = frame
+        mmap.flush()
+
+    # -------------------- private --------------------
+
+    def _set_sequence(self, seq: useq.MDASequence | None) -> None:
+        """Set the current sequence, and update the used axes."""
+        self._current_sequence = seq
+        if seq:
+            self._used_axes = tuple(seq.used_axes)
+
+    def _create_seq_memmap(
+        self, frame: np.ndarray, seq: useq.MDASequence, meta: dict
+    ) -> np.memmap:
+        from tifffile import imwrite, memmap
+
+        shape = (
+            *tuple(v for k, v in seq.sizes.items() if k in self._used_axes),
+            *frame.shape,
+        )
+        axes = (*self._used_axes, "y", "x")
+        dtype = frame.dtype
+
+        # see tifffile.tiffile for more metadata options
+        metadata: dict[str, Any] = {"axes": "".join(axes).upper()}
+        if seq:
+            if seq.time_plan and hasattr(seq.time_plan, "interval"):
+                interval = seq.time_plan.interval
+                if isinstance(interval, timedelta):
+                    interval = interval.total_seconds()
+                metadata["TimeIncrement"] = interval
+                metadata["TimeIncrementUnit"] = "s"
+            if seq.z_plan and hasattr(seq.z_plan, "step"):
+                metadata["PhysicalSizeZ"] = seq.z_plan.step
+                metadata["PhysicalSizeZUnit"] = "µm"
+            if seq.channels:
+                metadata["Channel"] = {"Name": [c.config for c in seq.channels]}
+        if acq_date := meta.get("Time"):
+            metadata["AcquisitionDate"] = acq_date
+        if pix := meta.get("PixelSizeUm"):
+            metadata["PhysicalSizeX"] = pix
+            metadata["PhysicalSizeY"] = pix
+            metadata["PhysicalSizeXUnit"] = "µm"
+            metadata["PhysicalSizeYUnit"] = "µm"
+
+        # TODO:
+        # there's a lot we could still capture, but it comes off the microscope
+        # over the course of the acquisition (such as stage positions, exposure times)
+        # ... one option is to accumulate these things and then use `tifffile.comment`
+        # to update the total metadata in sequenceFinished
```

Comment:
Yeah, this would also include real frame times for example, right? For the adaptive acquisitions these are pretty important and depending on the hardware, the TimeIncrement set is not always a very good measure for the real arrival time of frames. Is there a good value for this in the metadata received in frameReady?

Reply:
Yep exactly. Actual time would be present in the `meta` dict (third argument of frameReady) and would go into the Plane.delta_t field in the ome model.

Side note, see also the merged image writer in #267.  I imagine that one also being useful for you, since it would better handle possible ragged arrays that don't nicely fit into a clean nD model